File: WeekStory7/TechStackSwap.cpp
#include<iostream>
#include<vector>
#include<map>
#include<limits.h>
using namespace std;
class TechStackSwap {
    public:
    int solve(string& a, string& b, int p, int q, int ins, int del, int rep, vector<vector<int>>& memo) {
        int rows = a.size(), cols = b.size();
        if(p == rows && q == cols) {
            return 0;
        } else if(p == rows) {
            return (cols-q) * ins;
        } else if(q == cols) {
            return (rows-p) * del;
        }
        if(memo[p][q] != -1) {
            return memo[p][q];
        }
        if(a[p] == b[q]) {
            return memo[p][q] = solve(a, b, p+1, q+1, ins, del, rep, memo);
        } else {
            int add = solve(a, b, p, q+1, ins, del, rep, memo) + ins;
            int rem = solve(a, b, p+1, q, ins, del, rep, memo) + del;
            int swap = solve(a, b, p+1, q+1, ins, del, rep, memo) + rep;
            return memo[p][q] = min(add, min(rem, swap));
        }
    }
    int swapStacks() {
        string a, b;
        cout << "Enter tech two stacks:";
        getline(cin, a);
        getline(cin, b);
        int ins, del, rep;
        cout << "Enter insert, delete, replace cost:";
        cin >> ins >> del >> rep;
        int rows = a.size(), cols = b.size();
        vector<vector<int>> memo(rows, vector<int>(cols, -1));
        return solve(a, b, 0, 0, ins, del, rep, memo);
    }
};
int main() {
    TechStackSwap* obj = new TechStackSwap();
    cout << obj->swapStacks() << endl;
    return 0;
}

File: week_stories_7/TechStackSwap.java
import java.util.*;

public class TechStackSwap {

    public int minimumCost(String tech1, String tech2, int i, int j,
                           int insertCost, int deleteCost, int replaceCost,
                           int[][] dp) {

        int n = tech1.length();
        int m = tech2.length();

        if (i == n && j == m) {
            return 0;
        } 
        else if (i == n) {
            return (m - j) * insertCost;
        } 
        else if (j == m) {
            return (n - i) * deleteCost;
        }

        if (dp[i][j] != -1) {
            return dp[i][j];
        }

        if (tech1.charAt(i) == tech2.charAt(j)) {
            return dp[i][j] = minimumCost(
                    tech1, tech2, i + 1, j + 1,
                    insertCost, deleteCost, replaceCost, dp
            );
        } 
        else {

            int insert = minimumCost(
                    tech1, tech2, i, j + 1,
                    insertCost, deleteCost, replaceCost, dp
            ) + insertCost;

            int delete = minimumCost(
                    tech1, tech2, i + 1, j,
                    insertCost, deleteCost, replaceCost, dp
            ) + deleteCost;

            int replace = minimumCost(
                    tech1, tech2, i + 1, j + 1,
                    insertCost, deleteCost, replaceCost, dp
            ) + replaceCost;

            return dp[i][j] = Math.min(insert, Math.min(delete, replace));
        }
    }

    public int swapStacks() {

        Scanner sc = new Scanner(System.in);

        System.out.println("Enter tech two stacks:");
        String tech1 = sc.nextLine();
        String tech2 = sc.nextLine();

        System.out.println("Enter insert, delete, replace cost:");
        int insertCost = sc.nextInt();
        int deleteCost = sc.nextInt();
        int replaceCost = sc.nextInt();

        int n = tech1.length();
        int m = tech2.length();

        int[][] dp = new int[n][m];

        for (int[] row : dp) {
            Arrays.fill(row, -1);
        }

        return minimumCost(
                tech1, tech2, 0, 0,
                insertCost, deleteCost, replaceCost, dp
        );
    }

    public static void main(String[] args) {

        TechStackSwap techStacks = new TechStackSwap();

        System.out.println(techStacks.swapStacks());
    }
}
